# A walkthrough of text analysis and TF-IDF
#### Material from Google Colab Tutorials
#### Edited by Nakul Upadhya
We'll start by using scikit-learn to count words, then come across some of the issues with simple word count analysis. Most of these problems can be tackled with TF-IDF - a single word might mean less in a longer text, and common words may contribute less to meaning than more rare ones.

In [ ]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
import re
from nltk.stem.porter import PorterStemmer

pd.options.display.max_columns = 30
%matplotlib inline

## Text analysis refresher

Text analysis has a few parts. We are going to use **bag of words** analysis, which just treats a sentence like a bag of words - no particular order or anything. It's simple but it usually gets the job done adequately.

Here is our text.

In [ ]:
texts = [
    "Penny bought bright blue fishes.",
    "Penny bought bright blue and orange fish.",
    "The cat ate a fish at the store.",
    "Penny went to the store. Penny ate a bug. Penny saw a fish.",
    "It meowed once at the bug, it is still meowing at the bug and the fish",
    "The cat is at the fish store. The cat is orange. The cat is meowing at the fish.",
    "Penny is a fish"
]

When you process text, you have a nice long series of steps, but let's say you're interested in three things:

1. **Tokenizing** converts all of the sentences/phrases/etc into a series of words, and then it might also include converting it into a series of numbers - math stuff only works with numbers, not words. So maybe 'cat' is 2 and 'rug' is 4 and stuff like that.
2. **Counting** takes those words and sees how many there are (obviously) - how many times does `meow` appear?
3. **Normalizing** takes the count and makes new numbers - maybe it's how many times `meow` appears vs. how many total words there are, or maybe you're seeing how often `meow` comes up to see whether it's important.

In [ ]:
"Penny bought bright blue fishes".split()

['Penny', 'bought', 'bright', 'blue', 'fishes']

      Penny bought bright blue fishes.

If we **tokenize** that sentence, we're just lowercasing it, removing the punctuation and splitting on spaces - `penny bought bright blue fishes`.


The `scikit-learn` package does a **ton of stuff**, some of which includes the above. We're going to start by playing with the `CountVectorizer`.

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
count_vectorizer = CountVectorizer()

In [ ]:
# .fit_transfer TOKENIZES and COUNTS
X = count_vectorizer.fit_transform(texts)

Let's take a look at what it found out!

In [ ]:
X

<7x23 sparse matrix of type '<class 'numpy.int64'>'
	with 49 stored elements in Compressed Sparse Row format>

Okay, that looks like trash and garbage. What's a "sparse array"??????

In [ ]:
X.toarray()

array([[0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
        0],
       [1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0,
        0],
       [0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 0,
        0],
       [0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 3, 1, 0, 1, 1, 1,
        1],
       [1, 2, 0, 0, 0, 0, 2, 0, 1, 0, 1, 2, 1, 1, 1, 0, 0, 0, 1, 0, 3, 0,
        0],
       [0, 2, 0, 0, 0, 0, 0, 3, 2, 0, 3, 0, 0, 1, 0, 1, 0, 0, 0, 1, 5, 0,
        0],
       [0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
        0]])

If we put on our **Computer Goggles** we see that the first sentence has the first word 3 times, the second word 1 time, the third word 1 time, etc... But we can't *read* it, really. It would look nicer as a dataframe and with the words labeled.

In [ ]:
pd.DataFrame(X.toarray(), columns=count_vectorizer.get_feature_names_out())

and  at  ate  blue  bought  bright  bug  cat  fish  fishes  is  it  meowed  \
0    0   0    0     1       1       1    0    0     0       1   0   0       0   
1    1   0    0     1       1       1    0    0     1       0   0   0       0   
2    0   1    1     0       0       0    0    1     1       0   0   0       0   
3    0   0    1     0       0       0    1    0     1       0   0   0       0   
4    1   2    0     0       0       0    2    0     1       0   1   2       1   
5    0   2    0     0       0       0    0    3     2       0   3   0       0   
6    0   0    0     0       0       0    0    0     1       0   1   0       0   

   meowing  once  orange  penny  saw  still  store  the  to  went  
0        0     0       0      1    0      0      0    0   0     0  
1        0     0       1      1    0      0      0    0   0     0  
2        0     0       0      0    0      0      1    2   0     0  
3        0     0       0      3    1      0      1    1   1     1  
4        1     1       0      0    0      1      0    3   0     0  
5        1     0       1      0    0      0      1    5   0     0  
6        0     0       0      1    0      0      0    0   0     0

If we examine the resultant matrix, we can see that there are many words which are relatively redundant and may not be useful when it comes to distinguishing meaning of the sentence (ex. and, at, it, etc.). We call these **stopwords** and we can remove the stopwords from the matrix when we vectorize our points.

In [ ]:
# We'll make a new vectorizer
count_vectorizer = CountVectorizer(stop_words='english')
#count_vectorizer = CountVectorizer(stop_words=['the', 'and'])
# .fit_transfer TOKENIZES and COUNTS
X = count_vectorizer.fit_transform(texts)
print(count_vectorizer.get_feature_names_out())

['ate' 'blue' 'bought' 'bright' 'bug' 'cat' 'fish' 'fishes' 'meowed'
 'meowing' 'orange' 'penny' 'saw' 'store' 'went']


In [ ]:
pd.DataFrame(X.toarray(), columns=count_vectorizer.get_feature_names_out())

ate  blue  bought  bright  bug  cat  fish  fishes  meowed  meowing  orange  \
0    0     1       1       1    0    0     0       1       0        0       0   
1    0     1       1       1    0    0     1       0       0        0       1   
2    1     0       0       0    0    1     1       0       0        0       0   
3    1     0       0       0    1    0     1       0       0        0       0   
4    0     0       0       0    2    0     1       0       1        1       0   
5    0     0       0       0    0    3     2       0       0        1       1   
6    0     0       0       0    0    0     1       0       0        0       0   

   penny  saw  store  went  
0      1    0      0     0  
1      1    0      0     0  
2      0    0      1     0  
3      3    1      1     1  
4      0    0      0     0  
5      0    0      1     0  
6      1    0      0     0

I still see `meowed` and `meowing` and `fish` and `fishes` - they seem the same, so let's lemmatize/stem them.

You can specify a `preprocessor` or a `tokenizer` when you're creating your `CountVectorizer` to do custom *stuff* on your words. Maybe we want to get rid of punctuation, lowercase things and split them on spaces (this is basically the default). `preprocessor` is supposed to return a string, so it's a little easier to work with.

In [ ]:
# This is what our normal tokenizer looks like
def boring_tokenizer(str_input):
    words = re.sub(r"[^A-Za-z0-9\-]", " ", str_input).lower().split()
    return words

count_vectorizer = CountVectorizer(stop_words='english', tokenizer=boring_tokenizer)
X = count_vectorizer.fit_transform(texts)
print(count_vectorizer.get_feature_names_out())

['ate' 'blue' 'bought' 'bright' 'bug' 'cat' 'fish' 'fishes' 'meowed'
 'meowing' 'orange' 'penny' 'saw' 'store' 'went']


/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


We're going to use one that features a **stemmer** - something that strips the endings off of words (or tries to, at least). This one is from `nltk`.

In [ ]:
# https://tartarus.org/martin/PorterStemmer/def.txt
from nltk.stem.porter import PorterStemmer
porter_stemmer = PorterStemmer()

print(porter_stemmer.stem('fishes'))
print(porter_stemmer.stem('meowed'))
print(porter_stemmer.stem('oranges'))
print(porter_stemmer.stem('meowing'))
print(porter_stemmer.stem('orange'))
print(porter_stemmer.stem('go'))
print(porter_stemmer.stem('went'))

fish
meow
orang
meow
orang
go
went


In [ ]:
porter_stemmer = PorterStemmer()

def stemming_tokenizer(str_input):
    words = re.sub(r"[^A-Za-z0-9\-]", " ", str_input).lower().split()
    words = [porter_stemmer.stem(word) for word in words]
    return words

count_vectorizer = CountVectorizer(stop_words='english', tokenizer=stemming_tokenizer)
X = count_vectorizer.fit_transform(texts)
print(count_vectorizer.get_feature_names_out())

['ate' 'blue' 'bought' 'bright' 'bug' 'cat' 'fish' 'meow' 'onc' 'orang'
 'penni' 'saw' 'store' 'went']


/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:409: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['abov', 'afterward', 'alon', 'alreadi', 'alway', 'ani', 'anoth', 'anyon', 'anyth', 'anywher', 'becam', 'becaus', 'becom', 'befor', 'besid', 'cri', 'describ', 'dure', 'els', 'elsewher', 'empti', 'everi', 'everyon', 'everyth', 'everywher', 'fifti', 'formerli', 'forti', 'ha', 'henc', 'hereaft', 'herebi', 'hi', 'howev', 'hundr', 'inde', 'latterli', 'mani', 'meanwhil', 'moreov', 'mostli', 'nobodi', 'noon', 'noth', 'nowher', 'onc', 'onli', 'otherwis', 'ourselv', 'perhap', 'pleas', 'seriou', 'sever', 'sinc', 'sincer', 'sixti', 'someon', 'someth', 'sometim', 'somewher', 'themselv', 'thenc', 'thereaft', 'therebi', 'therefor', 'thi', 'thu', 'togeth', 'twelv', 'twenti', 'veri', 'wa', 'whatev', 'whenc', 'whenev', 'wherea', 'whereaft', 'wherebi', 'wherev', 'whi', 'yourselv'] not in stop_words.
  warnings

Now lets look at the new version of that dataframe.

In [ ]:
pd.DataFrame(X.toarray(), columns=count_vectorizer.get_feature_names_out())

ate      blue    bought    bright       bug    cat      fish  \
0  0.000000  0.200000  0.200000  0.200000  0.000000  0.000  0.200000   
1  0.000000  0.166667  0.166667  0.166667  0.000000  0.000  0.166667   
2  0.250000  0.000000  0.000000  0.000000  0.000000  0.250  0.250000   
3  0.111111  0.000000  0.000000  0.000000  0.111111  0.000  0.111111   
4  0.000000  0.000000  0.000000  0.000000  0.333333  0.000  0.166667   
5  0.000000  0.000000  0.000000  0.000000  0.000000  0.375  0.250000   
6  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.500000   

       meow       onc     orang     penni       saw     store      went  
0  0.000000  0.000000  0.000000  0.200000  0.000000  0.000000  0.000000  
1  0.000000  0.000000  0.166667  0.166667  0.000000  0.000000  0.000000  
2  0.000000  0.000000  0.000000  0.000000  0.000000  0.250000  0.000000  
3  0.000000  0.000000  0.000000  0.333333  0.111111  0.111111  0.111111  
4  0.333333  0.166667  0.000000  0.000000  0.000000  0.000000  0.000000  
5  0.125000  0.000000  0.125000  0.000000  0.000000  0.125000  0.000000  
6  0.000000  0.000000  0.000000  0.500000  0.000000  0.000000  0.000000

    "Penny bought bright blue fishes.",
    "Penny bought bright blue and orange fish.",
    "The cat ate a fish at the store.",
    "Penny went to the store. Penny ate a bug. Penny saw a fish.",
    "It meowed once at the bug, it is still meowing at the bug and the fish",
    "The cat is at the fish store. The cat is orange. The cat is meowing at the fish.",
    "Penny is a fish"

## TF-IDF

### Part One: Term Frequency

TF-IDF? What? It means **term frequency inverse document frequency!** It's the most important thing. Let's look at our list of phrases

1. Penny bought bright blue fishes.
2. Penny bought bright blue and orange fish.
3. The cat ate a fish at the store.
4. Penny went to the store. Penny ate a bug. Penny saw a fish.
5. It meowed once at the fish, it is still meowing at the fish. It meowed at the bug and the fish.
6. The cat is fat. The cat is orange. The cat is meowing at the fish.
7. Penny is a fish

If we are training a classifier to identify sentences about aquatic life, which is the most helpful phrase?

In [ ]:
pd.DataFrame(X.toarray(), columns=count_vectorizer.get_feature_names_out())

ate  blue  bought  bright  bug  cat  fish  meow  onc  orang  penni  saw  \
0    0     1       1       1    0    0     1     0    0      0      1    0   
1    0     1       1       1    0    0     1     0    0      1      1    0   
2    1     0       0       0    0    1     1     0    0      0      0    0   
3    1     0       0       0    1    0     1     0    0      0      3    1   
4    0     0       0       0    2    0     1     2    1      0      0    0   
5    0     0       0       0    0    3     2     1    0      1      0    0   
6    0     0       0       0    0    0     1     0    0      0      1    0   

   store  went  
0      0     0  
1      0     0  
2      1     0  
3      1     1  
4      0     0  
5      1     0  
6      0     0

Probably the one where `fish` appears two times.

    It meowed once at the fish, it is still meowing at the fish. It meowed at the bug and the fish.
    
But are all the others the same?

    Penny is a fish.

    Penny went to the store. Penny ate a bug. Penny saw a fish.

In the second one we spend less time talking about the fish. Think about a huge long document where they say your name once, versus a tweet where they say your name once. Which one are you more important in? Probably the tweet, since you take up a larger percentage of the text.

This is **term frequency** - taking into account how often a term shows up. We're going to take this into account by using the `TfidfVectorizer` in the same way we used the `CountVectorizer`.

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [ ]:
tfidf_vectorizer = TfidfVectorizer(stop_words='english', tokenizer=stemming_tokenizer, use_idf=False, norm='l1')
X = tfidf_vectorizer.fit_transform(texts)
pd.DataFrame(X.toarray(), columns=tfidf_vectorizer.get_feature_names_out())

/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:409: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['abov', 'afterward', 'alon', 'alreadi', 'alway', 'ani', 'anoth', 'anyon', 'anyth', 'anywher', 'becam', 'becaus', 'becom', 'befor', 'besid', 'cri', 'describ', 'dure', 'els', 'elsewher', 'empti', 'everi', 'everyon', 'everyth', 'everywher', 'fifti', 'formerli', 'forti', 'ha', 'henc', 'hereaft', 'herebi', 'hi', 'howev', 'hundr', 'inde', 'latterli', 'mani', 'meanwhil', 'moreov', 'mostli', 'nobodi', 'noon', 'noth', 'nowher', 'onc', 'onli', 'otherwis', 'ourselv', 'perhap', 'pleas', 'seriou', 'sever', 'sinc', 'sincer', 'sixti', 'someon', 'someth', 'sometim', 'somewher', 'themselv', 'thenc', 'thereaft', 'therebi', '

ate      blue    bought    bright       bug    cat      fish  \
0  0.000000  0.200000  0.200000  0.200000  0.000000  0.000  0.200000   
1  0.000000  0.166667  0.166667  0.166667  0.000000  0.000  0.166667   
2  0.250000  0.000000  0.000000  0.000000  0.000000  0.250  0.250000   
3  0.111111  0.000000  0.000000  0.000000  0.111111  0.000  0.111111   
4  0.000000  0.000000  0.000000  0.000000  0.333333  0.000  0.166667   
5  0.000000  0.000000  0.000000  0.000000  0.000000  0.375  0.250000   
6  0.000000  0.000000  0.000000  0.000000  0.000000  0.000  0.500000   

       meow       onc     orang     penni       saw     store      went  
0  0.000000  0.000000  0.000000  0.200000  0.000000  0.000000  0.000000  
1  0.000000  0.000000  0.166667  0.166667  0.000000  0.000000  0.000000  
2  0.000000  0.000000  0.000000  0.000000  0.000000  0.250000  0.000000  
3  0.000000  0.000000  0.000000  0.333333  0.111111  0.111111  0.111111  
4  0.333333  0.166667  0.000000  0.000000  0.000000  0.000000  0.000000  
5  0.125000  0.000000  0.125000  0.000000  0.000000  0.125000  0.000000  
6  0.000000  0.000000  0.000000  0.500000  0.000000  0.000000  0.000000

Now our numbers have shifted a little bit. Instead of just being a count, it's *the percentage of the words*.

    value = (number of times word appears in sentence) / (number of words in sentence)

After we remove the stopwords, the term `fish` is 50% of the words in `Penny is a fish` vs. 37.5% in `It meowed once at the fish, it is still meowing at the fish. It meowed at the bug and the fish.`.

> **Note:** We made it be the percentage of the words by passing in `norm="l1"` - by default it's normally an L2 (Euclidean) norm, which is actually better, but I thought it would make more sense using the L1 - a.k.a. terms divided by words -norm.

So now when we train a classifier, it will be able to identify the important words easier because our pre-processing takes into account whether half of our words are `fish` or 1% of millions upon millions of words is `fish`. But we aren't done yet!

### Part Two: Inverse document frequency

In addition to its importance to a given sentence, we also want to give information on how prevelant a given token is in the dataset as a whole. A token that is mentioned a lot in every dataset will probably be less important than a token that appears infrequently.

This is **inverse document frequency** - the more often a term shows up across *all* documents, the less important it is in our matrix.

In [ ]:
# use_idf=True is default, but I'll leave it in
idf_vectorizer = TfidfVectorizer(stop_words='english', tokenizer=stemming_tokenizer, use_idf=True, norm='l1')
X = idf_vectorizer.fit_transform(texts)
idf_df = pd.DataFrame(X.toarray(), columns=idf_vectorizer.get_feature_names_out())
idf_df

/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:409: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['abov', 'afterward', 'alon', 'alreadi', 'alway', 'ani', 'anoth', 'anyon', 'anyth', 'anywher', 'becam', 'becaus', 'becom', 'befor', 'besid', 'cri', 'describ', 'dure', 'els', 'elsewher', 'empti', 'everi', 'everyon', 'everyth', 'everywher', 'fifti', 'formerli', 'forti', 'ha', 'henc', 'hereaft', 'herebi', 'hi', 'howev', 'hundr', 'inde', 'latterli', 'mani', 'meanwhil', 'moreov', 'mostli', 'nobodi', 'noon', 'noth', 'nowher', 'onc', 'onli', 'otherwis', 'ourselv', 'perhap', 'pleas', 'seriou', 'sever', 'sinc', 'sincer', 'sixti', 'someon', 'someth', 'sometim', 'somewher', 'themselv', 'thenc', 'thereaft', 'therebi', '

ate      blue    bought    bright       bug       cat      fish  \
0  0.000000  0.235463  0.235463  0.235463  0.000000  0.000000  0.118871   
1  0.000000  0.190587  0.190587  0.190587  0.000000  0.000000  0.096216   
2  0.297654  0.000000  0.000000  0.000000  0.000000  0.297654  0.150267   
3  0.125073  0.000000  0.000000  0.000000  0.125073  0.000000  0.063142   
4  0.000000  0.000000  0.000000  0.000000  0.350291  0.000000  0.088420   
5  0.000000  0.000000  0.000000  0.000000  0.000000  0.437035  0.147088   
6  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.404858   

       meow       onc     orang     penni       saw     store      went  
0  0.000000  0.000000  0.000000  0.174741  0.000000  0.000000  0.000000  
1  0.000000  0.000000  0.190587  0.141437  0.000000  0.000000  0.000000  
2  0.000000  0.000000  0.000000  0.000000  0.000000  0.254425  0.000000  
3  0.000000  0.000000  0.000000  0.278455  0.150675  0.106908  0.150675  
4  0.350291  0.210997  0.000000  0.000000  0.000000  0.000000  0.000000  
5  0.145678  0.000000  0.145678  0.000000  0.000000  0.124521  0.000000  
6  0.000000  0.000000  0.000000  0.595142  0.000000  0.000000  0.000000

Notice how 'meow' increased in value because it's an infrequent term, and `fish` dropped in value because it's so frequent.

That meowing one (index 4) has gone from `0.50` to `0.43`, while `Penny is a fish` (index 6) has dropped to `0.40`. Now hooray, the meowing one is going to show up earlier when searching for "fish meow" because *fish shows up all of the time, so we want to ignore it a lil' bit*.

Let's try changing it to `norm='l2'` (or just removing `norm` completely) to see how this might change our matrix.

In [ ]:
# use_idf=True is default, but I'll leave it in
l2_vectorizer = TfidfVectorizer(stop_words='english', tokenizer=stemming_tokenizer, use_idf=True)
X = l2_vectorizer.fit_transform(texts)
l2_df = pd.DataFrame(X.toarray(), columns=l2_vectorizer.get_feature_names_out())
l2_df

/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:409: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['abov', 'afterward', 'alon', 'alreadi', 'alway', 'ani', 'anoth', 'anyon', 'anyth', 'anywher', 'becam', 'becaus', 'becom', 'befor', 'besid', 'cri', 'describ', 'dure', 'els', 'elsewher', 'empti', 'everi', 'everyon', 'everyth', 'everywher', 'fifti', 'formerli', 'forti', 'ha', 'henc', 'hereaft', 'herebi', 'hi', 'howev', 'hundr', 'inde', 'latterli', 'mani', 'meanwhil', 'moreov', 'mostli', 'nobodi', 'noon', 'noth', 'nowher', 'onc', 'onli', 'otherwis', 'ourselv', 'perhap', 'pleas', 'seriou', 'sever', 'sinc', 'sincer', 'sixti', 'someon', 'someth', 'sometim', 'somewher', 'themselv', 'thenc', 'thereaft', 'therebi', '

ate      blue    bought    bright       bug       cat      fish  \
0  0.000000  0.512612  0.512612  0.512612  0.000000  0.000000  0.258786   
1  0.000000  0.456170  0.456170  0.456170  0.000000  0.000000  0.230292   
2  0.578752  0.000000  0.000000  0.000000  0.000000  0.578752  0.292176   
3  0.303663  0.000000  0.000000  0.000000  0.303663  0.000000  0.153301   
4  0.000000  0.000000  0.000000  0.000000  0.641958  0.000000  0.162043   
5  0.000000  0.000000  0.000000  0.000000  0.000000  0.840166  0.282766   
6  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.562463   

       meow       onc     orang     penni       saw     store      went  
0  0.000000  0.000000  0.000000  0.380417  0.000000  0.000000  0.000000  
1  0.000000  0.000000  0.456170  0.338530  0.000000  0.000000  0.000000  
2  0.000000  0.000000  0.000000  0.000000  0.000000  0.494698  0.000000  
3  0.000000  0.000000  0.000000  0.676058  0.365821  0.259561  0.365821  
4  0.641958  0.386682  0.000000  0.000000  0.000000  0.000000  0.000000  
5  0.280055  0.000000  0.280055  0.000000  0.000000  0.239382  0.000000  
6  0.000000  0.000000  0.000000  0.826823  0.000000  0.000000  0.000000

## Real Life Example

Lets take what we learned and apply it to a real life classification task. AG is a collection of more than 1 million news articles organized into four different topics: Business, Tech, Sports, and World News.
Your task is to:
1. Apply the text pre-processing you learnt above to get a clean dataset.
2. Train a classifier on your processed data.
3. Report the training and testing error.

Code to download and sample the data is already provided.



In [ ]:
!pip install datasets

In [41]:
from datasets import load_dataset

dataset = load_dataset("ag_news", cache_dir='data/datasets/text/ag_news')
train = dataset['train']
test = dataset['test']
# convert to pandas
train = pd.DataFrame(train).head(2000)
test = pd.DataFrame(test).head(500)
print(train.shape)
print(test.shape)
train.head()

(2000, 2)
(500, 2)


In [ ]:
## Your code here